쥬스메이커 [STEP 2] Gundy, 준호

junho15 opened the pull request:
@junbangg
안녕하세요 알라딘!🏜
두 번째 PR도 잘 부탁드립니다.
약간 두서없이 적은 내용도 있지만...우리의 프린스 알리는 지니처럼 찰떡같이 알아들을 것이라고 믿습니다! 감사해요!!🙌

#### 함수별 기능
- `showOkayAlert`: 쥬스가 완성되었을 때 호출할 얼럿을 띄우는 메서드
- `showStockEditAlert`: 주문받은 쥬스를 만들 과일의 재고가 부족할 때, 재고 수정 페이지로 넘어갈 것인지 선택을 받는 얼럿을 띄우는 메서드
- `touchUpJuiceOrderButton`: 각 쥬스별로 주문 버튼을 터치했을 때 쥬스를 만들게 하는 메서드
- `presentStockEditView`: 재고 수정 뷰로 화면전환을 하는 메서드
- `touchUpDismissButton`: 닫기 버튼을 누르면 재고수정 뷰에서 빠져나오게 하는 메서드
- `updateFruitStockLabel`: 과일의 재고 데이터를 뷰의 레이블에 업데이트하는 메서드

### 코드를 작성할 때 고민되었던 점
- FruitStoreError를 캐치해 print하던 `make` 메서드를 이용해 쥬스를 만드려다보니, 반환 값이 필요해졌습니다. 그 반환에 따라 어느 얼럿을 띄울 것인지를 선택해야 했기 때문입니다. 기존에 `make` 메서드를 작성할 때도 오류처리를 한 것은 나름의 이유가 있었기 때문에 이 상황을 쉽게 해결하려고 오류처리를 없앨 수는 없었습니다. 그 결과 반환 타입을 `Result` 타입으로 선택해 적절한 얼럿을 호출하게 하였습니다.
- 이번 스텝의 구현 사항에는 재고수정 뷰로 화면전환까지만 구현할 뿐, 그 화면에서 빠져나오는 기능을 요구하지 않습니다. 하지만 테스트 과정이나 리뷰의 편의를 생각해 닫기 버튼을 구현하는 게 옳다고 생각하였습니다. 다음 스텝에서 요구하는 버튼과는 달라졌지만, 수정을 할 생각으로 만들었습니다.

### 조언을 얻고 싶은 부분
- 얼럿에 들어갈 메시지를 문자열로 전달할 때나 얼럿 액션의 title 같은 경우는 현재 매직리터럴로 보입니다. 모든 경우에서 매직리터럴은 배열 등으로 따로 타입을 정의해서 처리해 주는 것이 좋을까요?

- `updateFruitStockLabel` 메서드를 작성하면서, 새로 오류 처리 없이 과일의 재고를 반환하는 메서드를 생성하기보다는 기존에 구현했던 메서드인 `currentStock`을 호출해 사용하기로 결정하였습니다. 다만 `currentStock` 메서드는 원래 `checkInventoryHasStock` 메서드에서 현재 창고 품목에 있는지 확인하고 수량을 가져오는 용도로 호출해 사용하고 있었는데, `checkInventoryHasStock` 메서드와 `currentStock`메서드를 분리하는 것이 더 좋았을까요? `updateFruitStockLabel` 메서드에서는 던져진 오류에 적절한 처리를 해주지 않기 때문에 이런 고민이 들었습니다.
- if let 문에 else 블록을 추가해서 에러가 발생한 경우를 추가하는 게 좋을까요?
~~~swift
    func updateFruitStockLabel() {
        if let strawberryStock = try? juiceMaker.fruitStore.currentStock(of: .strawberry) {
            strawberryStockLabel.text = "\(strawberryStock)"
        }
        if let bananaStock = try? juiceMaker.fruitStore.currentStock(of: .banana) {
            bananaStockLabel.text = "\(bananaStock)"
        }
        if let kiwiStock = try? juiceMaker.fruitStore.currentStock(of: .kiwi) {
            kiwiStockLabel.text = "\(kiwiStock)"
        }
        if let pineappleStock = try? juiceMaker.fruitStore.currentStock(of: .pineapple) {
            pineappleStockLabel.text = "\(pineappleStock)"
        }
        if let mangoStock = try? juiceMaker.fruitStore.currentStock(of: .mango) {
            mangoStockLabel.text = "\(mangoStock)"
        }
    }
~~~

- fruitStore 상수를 만든 뒤에 JuiceMaker 이니셜라이저에 만든 fruitStore를 전달해서 저장 프로퍼티에 할당하려고 했는 데 아래와 같은 에러가 발생했습니다.
- 아래 사용한 방법처럼 이니셜라이저에서 바로 FruitStore의 인스턴스를 만들어서 전달하니 에러는 발생하지 않았습니다.
- 왜 두 가지 방법의 결과가 다른 지 궁금합니다.

- 알아보니 ViewController의 인스턴스가 만들어 지기 전에는 self의 프로퍼티에 접근할 수 없어, fruitStore 프로퍼티를 사용할 수 없기 때문에 juiceMaker의 기본값을 초기화 할 때 fruitStore 프로퍼티를 사용할 수 없었습니다. lazy를 붙여 지연 저장 프로퍼티로 하면 사용할 때 초기화가 되기 때문에 fruitStore 프로퍼티를 사용할 수 있습니다. 메서드는 프로퍼티와 달리 인스턴스 생성 후에 사용되는 것이므로 self의 프로퍼티에 대한 접근이 가능했습니다. 뷰 컨트롤러의 인스턴스를 이니셜라이저를 통해 만들어주면 Lazy 키워드 없이도 지금처럼 juiceMaker와 fruitStore를 사용가능한지 궁금합니다.
- ViewController의 인스턴스화는 어디서 어떻게 되는지도 궁금합니다.
~~~swift
    // 시도했던 방법 - 실패
    let fruitStore: FruitStore = FruitStore(initialStock: 10)
    var juiceMaker: JuiceMaker = .init(fruitStore: fruitStore) 
    // Cannot use instance member 'fruitStore' within property initializer; property initializers run before 'self' is available
    
    /*
    struct JuiceMaker {
    let fruitStore: FruitStore
    
    init(fruitStore: FruitStore) {
      self.fruitStore = fruitStore
    }
    */

    // 사용한 방법
    var juiceMaker: JuiceMaker = .init(fruitStore: .init(initialStock: 10))

    // 수정한 방법
    var fruitStore: FruitStore = .init(initialStock: 10)
    lazy var juiceMaker: JuiceMaker = .init(fruitStore: fruitStore)
~~~

- 쥬스 제조 후 완성 메시지를 출력하는 showOkayAlert 메서드와 재고가 부족하다는 메시지를 출력하는 showStockEditAlert 메서드를 만들었습니다.
- showOkayAlert 메서드는 출력 메시지를 다르게하면 재사용이 가능해서 showOkayAlert로 네이밍을 했습니다.
- showStockEditAlert 메서드는 재고가 부족하다는 메시지를 출력한 뒤 사용자가 예/아니오 중 무엇을 선택하는 지에 따라 재고수정화면으로 이동하거나 얼럿이 닫힙니다. 특정 행위(재고수정화면으로 이동)가 바로 이어서 발생할 수 있기 때문에 ShowStockEditAlert로 네이밍을 했습니다.
- 네이밍이 적절한 지 궁금합니다.
~~~swift
    func showOkayAlert(_ message: String) {
        let alert = UIAlertController(title: nil,
                                      message: message,
                                      preferredStyle: .alert)
        let okAction = UIAlertAction(title: "OK",
                                     style: .default,
                                     handler: nil)
        alert.addAction(okAction)
        present(alert,
                animated: true,
                completion: nil)
    }
    
    func showStockEditAlert(_ message: String) {
        let alert = UIAlertController(title: nil,
                                      message: message,
                                      preferredStyle: .alert)
        let editAction = UIAlertAction(title: "예",
                                       style: .default) { (action) in
            self.prsentStockEditView()
        }
        let cancelAction = UIAlertAction(title: "아니오",
                                         style: .default)
        alert.addAction(editAction)
        alert.addAction(cancelAction)
        present(alert,
                animated: true,
                completion: nil)
    }
~~~

- viewController의 메서드는 은닉화를 어떻게 하면 좋을 지 궁금합니다.과연 뷰 컨트롤러에서는 은닉화가 의미가 있는가? 이니셜라이저가 없는데 뷰 인스턴스는 생성이 되는가? 하는 생각이 들었습니다. 이에 조언을 구하고 싶습니다.

junho15 committed:
feat: 재고수정 뷰 컨트롤러 추가, 재고 추가 화면 관련 화면 전환 기능 추가

Gundy93 committed:
feat: 쥬스 완성시 얼럿 메시지를 표시하는 showOKAlert 메서드 추가

junho15 committed:
feat: 재고 부족 시 얼럿 메시지를 표시하는 showEditAlert 메서드 추가

Gundy93 committed:
feat: 각 주문 버튼 터치시 쥬스를 만드는 touchUpJuiceOrderButton 메서드 추가

- 열거형 FruitStoreError에 unexpectedError 케이스 추가
- 열거형 Juice에 쥬스의 이름을 반환하는 연산 프로퍼티 name 추가
- 구조체 JuiceMaker의 make 메서드 수정

junho15 committed:
feat: 재고 추가 화면 관련 화면 전환하는 presentStockEditView 메서드 추가

- 기존 StoryBoard 상의 기능 대체

Gundy93 committed:
feat: 화면의 재고 레이블의 데이터를 변경하는 updateFruitStockLabel 메서드 및 UILabel 추가

junho15 committed:
refactor: showOkAlert 메서드 showOkayAlert로 이름 변경, showEditAlert 메서드 showStockEditAlert로 이름 변경

Gundy93 committed:
refactor: ViewController의 juiceMaker를 Lazy로 변경, fruitStore 프로퍼티 추가

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,11 +7,118 @@
+    @IBOutlet weak var bananaStockLabel: UILabel!
+    @IBOutlet weak var kiwiStockLabel: UILabel!
+    @IBOutlet weak var pineappleStockLabel: UILabel!
+    @IBOutlet weak var mangoStockLabel: UILabel!
```
`ViewController` 에 은닉화가 필요하냐는 질문을 해주셨는데
두분은 은닉화가 필요한 이유가 무엇이라고 생각하는지 궁금합니다!

Gundy93 replied:
저는 사용자가 직접적으로 컨트롤하지 않거나, 보여주고 싶지 않은 부분에 대해서 숨기는 것이라고 생각합니다. 또한 특정 데이터를 적절하지 않게 변경하지 않도록 취하는 조치라고도 생각합니다. 이로 인해 은닉화 된 기능에 접근할 수 있는 방법을 제시하는 것이 캡슐화라고 생각합니다.

junbangg replied:
말씀하신 부분들에 대해서 전부 동의합니다👍👍
그런데 추가적으로 저는 은닉화도 일종의 컨벤션으로도 여길 수 있다고 생각해요
코드베이스가 방대해질수록 코드는 점점 이해하기 힘들어지는데 은닉화가 갑자기 안되어있는 코드가 있으면 혼란이 생길 수 있을것 같아요.
어? 얘 밖에서 접근할 일도 없을텐데 왜 접근제어가 안되어있지??🤔 하고 말이죠

그래서 저는 객체지향의 기본 원칙 중의 하나인 "은닉화" 를 코드에서 지키기로 했으면, "통일성" 이라는 근건로도 충분히 모든 타입에 대해서 적용시키는것도 나쁘지 않다고 봅니다.
(요건 "뷰컨에서의 은닉화" 관련해서 질문 주신거에 대한 개인적인 생각이었습니다.)

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,11 +7,118 @@
+    @IBAction func touchUpJuiceOrderButton(_ sender: UIButton) {
+        let juice: Juice
+        switch sender {
+        case strawberryJuiceOrderButton:
```
👍

junbangg replied:
요 메서드의 덩치를 좀 줄여볼 수는 없을까요??

Gundy93 replied:
b25c8c1 커밋에서 메서드 분리를 통해 덩치를 줄였습니다!

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,11 +7,118 @@
+                completion: nil)
+    }
+    
+    func prsentStockEditView() {
```
오타가 있어보이네요!

Gundy93 replied:
`presentStockEditView`로 수정하겠습니다!

Gundy93 replied:
7e469e9 커밋에서 오타 수정했습니다

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,11 +7,118 @@
+        }
+        if let mangoStock = try? fruitStore.currentStock(of: .mango) {
+            mangoStockLabel.text = "\(mangoStock)"
+        }
```
여기서 에러를 반환하는것에 대한 질문을 하셨죠????
이전에도 여쭤봤었지만 에러 핸들링을 하는 이유에 대해서 한번 생각해보면 좋을것 같습니다 
지금 사용하고 계신 이유가 뭔가요?! (요구사항에 있었기 때문에는 제외ㅎㅎ👍)

Gundy93 replied:
프로그램이 개발자가 의도한대로 동작할 것이란 보장은 없기 때문에 의도하지 않은 동작이 발생하면 정상화 작업을 해주기 위해서 오류처리가 필요한 것 같습니다. 더불어 내부 코드를 알지 못하는 사용자에게 오류에 대한 내용을 전달하는 역할도 할 것 같습니다.

junbangg replied:
저도 같은 생각입니다 👍👍

junbangg commented on JuiceMaker/JuiceMaker/Model/FruitStore.swift:
```diff
@@ -62,7 +62,7 @@ class FruitStore {
     }
     
-    func getCurrentStock(of fruit: Fruit) throws -> Int {
+    func currentStock(of fruit: Fruit) throws -> Int {
```
요 메서드명 변경의 이유는 무엇일까요?!

Gundy93 replied:
네이밍을 더 잘하고 싶어서 네이밍 컨벤션을 다시 봤습니다. 
Name functions and methods according to their side-effects
• Those without side-effects should read as noun phrases, e.g. x.distance(to: y), i.successor().
-Swift API Design Guidelines - Naming - Strive for Fluent Usage
이 파트의 내용으로 미루어보아 get이라는 단어가 빠지는 것이 더 적절할 것이라고 생각했습니다.

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,11 +7,118 @@
+    }
+    
+    func showStockEditAlert(_ message: String) {
+        let alert = UIAlertController(title: nil,
```
지금 `UIAlertController` 가 위에서도 한번, 여기서도 한번 생성되고 있네요
재사용할 수 있는 방법은 없을까요??

Gundy93 replied:
7e469e9 커밋에서 가변 매개변수로 UIAlertAction을 전달 받아 Alert을 띄우는 `showAlert` 메서드 추가해서 재사용을 방지했습니다.

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,11 +7,118 @@
+        default:
+            return
+        }
+        let result: Result<Juice,FruitStoreError> = juiceMaker.make(juice)
```
`Juice,FruitStoreError` 사이에 띄어쓰기가 필요해보입니다!

Gundy93 replied:
b25c8c1 커밋에서 띄어쓰기 수정했습니다!

junbangg commented on JuiceMaker/JuiceMaker/Controller/ViewController.swift:
```diff
@@ -7,11 +7,118 @@
+        case .failure:
+            showStockEditAlert("재료가 모자라요. 재고를 수정할까요?")
+        }
+    }
```
음.. `make`의 반환값이 필요해서 `Result`를 사용해주셨는데 .. 반환값을 전혀 사용하지 않고 있는것으로 보입니다..!

지금 `make` 에 들어가기 전의 `juice` 를  `.success` 일때 쓰고 있고 (`Result`안에 반환된 `juice` 가 있을텐데 말이죠)
`.failure` 일때는 `make` 에서 나오는 오류들을 전혀 사용하지 않고 있는것으로 보여요.

그래서 지금은 `Result` 를 사용하는 근거/이점이 없어보입니다ㅜ

"make 메서드를 이용해 쥬스를 만드려다보니, 반환 값이 필요해졌습니다. 그 반환에 따라 어느 얼럿을 띄울 것인지를 선택해야 했기 때문입니다."

단순히 딱 `성공`,`실패` 를 구분해야되는거라면 `Bool` 반환으로도 충분히 가능할것 같은데 말이죠...
지금은 `Result` 가 `Bool`, `Optional` 이상의 역할을 못해주는것 같습니다.
조금 더 `Result` 사용법과 맞게 리팩토링을 하거나, 다른 방법을 찾는게 좋아보입니다!

이 모든 이야기들은 결국 우리가 오류제어를 왜 하는지에 대한 이야기로 다시 돌아가는것 같습니다. :)

Gundy93 replied:
0b49310 커밋에서 오류 처리 관련 수정을 거쳤습니다. 성공한 경우 반환되는 juice를 사용하는 것을 명시하기 위해 `madeJuice`로 네이밍을 해 사용하고 있습니다. 또한 실패한 경우 해당 오류에 맞는 얼럿을 불러와 오류처리를 하는데 현재는 재고부족 외에는 그냥 얼럿에 메시지만 담아 띄우지만, STEP-3에서 재고 추가,감소하는 기능을 구현할 때 notInInventoryFruitList 관련 수정 얼럿도 추가될 것으로 예상됩니다.

junbangg commented on JuiceMaker/JuiceMaker/Model/JuiceMaker.swift:
```diff
@@ -4,25 +4,30 @@
-    func make(_ juice: Juice) {
+    func make(_ juice: Juice) -> Result<Juice, FruitStoreError> {
         do {
             try fruitStore.checkStockOfIngredients(in: juice.recipe)
```
지금 `FruitStore`에서부터 열심히 전달만 되고 있는 오류들이 빛을 못보고 있는것 같아서 마음이 아픕니다🥲

Gundy93 replied:
오류에 관한 것을 더 고민해보아야 관련 코멘트들에 대답을 잘 할 수 있을 것 같습니다...🥲 더 고민해보고 알아보겠습니다.

Gundy93 replied:
0b49310 커밋에서 make로부터 전달되는 오류에 맞게 수정했습니다.

junbangg reviewed (changes requested):
이전 스텝도 수고 많으셨습니다!
고민의 흔적도 보이고 네이밍도 신경 많이 쓰신것 같아요!! 👍 
댓글 참고해주시면 될것 같고 질문 주신 내용 답변 드리겠습니다.

## 조언을 얻고 싶은 부분
```
"얼럿에 들어갈 메시지를 문자열로 전달할 때나 얼럿 액션의 title 같은 경우는 현재 매직리터럴로 보입니다. 모든 경우에서 매직리터럴은 배열 등으로 따로 타입을 정의해서 처리해 주는 것이 좋을까요?"
```
말씀하신것 처럼 `Namespace` 를 활용해서 매직 리터럴들을 모아두면 생기는 이점에는 어떤것들이 있다고 생각하시나요???

```
"updateFruitStockLabel 메서드를 작성하면서, 새로 오류 처리 없이 과일의 재고를 반환하는 메서드를 생성하기보다는 기존에 구현했던 메서드인 currentStock을 호출해 사용하기로 결정하였습니다. 다만 currentStock 메서드는 원래 checkInventoryHasStock 메서드에서 현재 창고 품목에 있는지 확인하고 수량을 가져오는 용도로 호출해 사용하고 있었는데, checkInventoryHasStock 메서드와 currentStock메서드를 분리하는 것이 더 좋았을까요? updateFruitStockLabel 메서드에서는 던져진 오류에 적절한 처리를 해주지 않기 때문에 이런 고민이 들었습니다."
```
메서드를 재사용하는것은 좋아보입니다. 
근데 해당 메서드를 `checkInventoryHasStock` 메서드에서 사용했다고 해서 재사용이 안되는 이유가 잘 이해 안됩니다! 질문을 제대로 이해 못한것 같네요 제가 ㅜ

```
"if let 문에 else 블록을 추가해서 에러가 발생한 경우를 추가하는 게 좋을까요?"
```
댓글에도 여러번 이야기 했지만 두 분은 오류처리를 하는 이유가 뭐라고 생각하시는지 궁금합니다:)

```
뷰컨트롤러의 생성에 대해서
```
말씀하신것처럼 저장 프로퍼티들이 다 정해지기 전에는 `self` 를 사용할 수 없습니다. 말그대로 인스턴스화가 덜 된것이니까요.
그래서 말씀해주신것처럼 그럴때는 지연 프로퍼티를 활용할 수도 있습니다.

그리고 뷰컨도 생성하는 방법이 있습니다. 
앱실행의 순서를 공부해보면 알 수 있지 않을까 싶습니다:) 

첫번째 뷰컨은 그렇고, 만약에 두번째 뷰컨 (`StockEditViewController`) 를 생성하고 싶으면 
`func instantiateViewController(withIdentifier: , creator: )` 를 알아봐주시면 좋을것 같습니다~~

이 이야기들은 나중에 공부하시겠지만 
- `ViewController 의 생명주기`
- `App 의 생명주기`
공부하시면 좀 풀릴 고민들로 보입니다.

더 궁금하신거 있으시면 댓글 남겨주세요~

수고하셨습니다~

junho15 committed:
refactor: 버튼에서 주문하는 주스가 무엇인지 반환하는 juice 메서드 추가해서 touchUpJuiceOrderButton 메서드에서 분리, FruitStoreEror 프로퍼티 failureReason을 localizedDescription으로 변경

Gundy93 committed:
refactor: 기존 매직 리터럴을 담을 AlertText 열거형 추가

junho15 committed:
refactor: 메시지와 가변 매개변수로 UIAlertAction을 전달 받아 Alert을 띄우는 showAlert 메서드 추가해서 showOkayAlert와 showStockEditAlert에서 활용

Gundy93 committed:
refactor: 오류처리 관련 수정

- 수정사항이 시뮬레이터 상에서 잘 보이도록 파인애플 재고 없이 초기화
- FruitStoreError의 notInInventoryFruitList 일 때 재고의 nil을 표현하기 위한 notExist 프로퍼티 추가
- JuiceMaker의 make 메서드에서 catch 하지 않는 invalidAmount 에러 캐치구문 삭제
- make의 Result 중 failure일 때, 오류별로 오류에 맞는 얼럿 메서드 호출

Gundy93 commented:
> 말씀하신것 처럼 Namespace 를 활용해서 매직 리터럴들을 모아두면 생기는 이점에는 어떤것들이 있다고 생각하시나요???

코드 작성자가 아닌 사람이 코드를 볼 때 코드의 흐름을 더욱 자연스럽게 이해하게 하고, 리팩토링을 거칠 때도 코드 전체가 한번에 수정되기 때문에 수정 또한 용이하다고 생각합니다.

> ViewController 의 생명주기
App 의 생명주기
공부하시면 좀 풀릴 고민들로 보입니다.

이 부분은 더욱 공부해보겠습니다! 현재 더 궁금한 점은 공부하다가 생기면 여쭤보겠습니다!!

junbangg commented:
> > 말씀하신것 처럼 Namespace 를 활용해서 매직 리터럴들을 모아두면 생기는 이점에는 어떤것들이 있다고 생각하시나요???
> 
> 코드 작성자가 아닌 사람이 코드를 볼 때 코드의 흐름을 더욱 자연스럽게 이해하게 하고, 리팩토링을 거칠 때도 코드 전체가 한번에 수정되기 때문에 수정 또한 용이하다고 생각합니다.
> 
> > ViewController 의 생명주기
> > App 의 생명주기
> > 공부하시면 좀 풀릴 고민들로 보입니다.
> 
> 이 부분은 더욱 공부해보겠습니다! 현재 더 궁금한 점은 공부하다가 생기면 여쭤보겠습니다!!

`Namespace` 를 사용하기 위한 충분한 근거가 될것 같습니다~
추가로 여러번 사용되는 문자열에 대한 중복적인 문자열 생성도 방지할 수 있을것 같다는 생각도 드네요👍👍
